Add backpressure to the unsigned extrinsic submission in executor

liuchengxu opened the pull request:
This PR aims to support submitting the unsigned extrinsic with some backpressure. Given that these unsigned extrinsics are kind of crucial, it's better not to drop any of them, we choose to slow down the message producers, which are the bundle processor and the gossiped executor message handler. Whenever the message producers are invoked, they'll first check if there is a pending message and try to resend it if there is one, if the resend is successful, continue as normal, otherwise, skip the further processing.

The buffer size is hardcoded to 128 for now, can be tweaked once we have a better idea.

### Code contributor checklist:
* [x] I have reviewed my own changes one more time to spot typos, unintended changes, following project conventions, etc.
* [x] I have prepared clean readable history of commits before submitting this PR to make reviewer's life easier
* [x] I have tested my changes and/or added corresponding test cases (if relevant)
* [x] I understand that any changes to this PR going forward will notify multiple developers and will try to minimize them
* [x] I have added sufficient description of changes to make review process easier
* [x] This PR is ready for review by developers

liuchengxu committed:
Add a new module `unsigned_submitter` in cirrus-executor

liuchengxu committed:
Integrate UnsignedSubmitter into bundle processor

While processing the bundles, first check if there is a pending message
and try sending via `try_submit()` again if there is one, if `try_submit()`
is successful, continue the bundles processing as normal, otherwise,
skip the bundles processing as there are too many unsigned proof
messages waiting to be processed.

liuchengxu committed:
Integrate UnsignedSubmitter to the gossiped message handler

Skip processing the gossiped Bundle and Receipt if there are too many
unsigned proof messages waiting to be processed.

nazar-pc commented on cumulus/client/cirrus-executor/src/unsigned_submitter.rs:
```diff
@@ -0,0 +1,111 @@
+        Block: BlockT,
+        PBlock: BlockT,
+        PClient: HeaderBackend<PBlock> + ProvideRuntimeApi<PBlock> + 'static,
+        PClient::Api: ExecutorApi<PBlock, Block::Hash>,
```
Doesn't this work too?:
```suggestion

    pub(crate) fn new<Block, PBlock, PClient, Spawner>(
        primary_chain_client: Arc<PClient>,
        spawner: &Spawner,
    ) -> Self
    where
        Block: BlockT,
        PBlock: BlockT,
        PClient: HeaderBackend<PBlock> + ProvideRuntimeApi<PBlock> + 'static,
        PClient::Api: ExecutorApi<PBlock, Block::Hash>, 
        Spawner: SpawnNamed,
```

It doesn't seem to be necessary to move it anywhere or own.

liuchengxu replied:
This does not compile unfortunately, but owning it is unnecessary indeed, this works `&Box<dyn SpawnNamed + Send + Sync>`.

```bash
error[E0277]: the trait bound `Box<dyn SpawnNamed + Sync + std::marker::Send>: SpawnNamed` is not satisfied
   --> cumulus/client/cirrus-executor/src/lib.rs:234:13
    |
232 |         let unsigned_submitter = UnsignedSubmitter::new::<Block, PBlock, PClient, _>(
    |                                  --------------------------------------------------- required by a bound introduced by this call
233 |             primary_chain_client.clone(),
234 |             &spawner,
    |             ^^^^^^^^ the trait `SpawnNamed` is not implemented for `Box<dyn SpawnNamed + Sync + std::marker::Send>`
    |
    = help: the trait `SpawnNamed` is implemented for `Box<(dyn SpawnNamed + 'static)>`
note: required by a bound in `UnsignedSubmitter::new`
   --> cumulus/client/cirrus-executor/src/unsigned_submitter.rs:39:18
    |
30  |     pub(crate) fn new<Block, PBlock, PClient, Spawner>(
    |                   --- required by a bound in this
...
39  |         Spawner: SpawnNamed
    |                  ^^^^^^^^^^ required by this bound in `UnsignedSubmitter::new`

For more information about this error, try `rustc --explain E0277`.
```

nazar-pc replied:
I didn't try to compile it, but this simple change compiles:
```patch
Index: cumulus/client/cirrus-executor/src/lib.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/cumulus/client/cirrus-executor/src/lib.rs b/cumulus/client/cirrus-executor/src/lib.rs
--- a/cumulus/client/cirrus-executor/src/lib.rs	(revision 9d6e45856c2f86fe9bd522b7ebbe9bbdc9dd0e1c)
+++ b/cumulus/client/cirrus-executor/src/lib.rs	(date 1658337149847)
@@ -229,9 +229,9 @@
             code_executor,
         );
 
-        let unsigned_submitter = UnsignedSubmitter::new::<Block, PBlock, PClient>(
+        let unsigned_submitter = UnsignedSubmitter::new::<Block, PBlock, PClient, _>(
             primary_chain_client.clone(),
-            spawner.clone(),
+            spawner.as_ref(),
         );
 
         let bundle_processor = BundleProcessor::new(
Index: cumulus/client/cirrus-executor/src/unsigned_submitter.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/cumulus/client/cirrus-executor/src/unsigned_submitter.rs b/cumulus/client/cirrus-executor/src/unsigned_submitter.rs
--- a/cumulus/client/cirrus-executor/src/unsigned_submitter.rs	(revision 9d6e45856c2f86fe9bd522b7ebbe9bbdc9dd0e1c)
+++ b/cumulus/client/cirrus-executor/src/unsigned_submitter.rs	(date 1658337109536)
@@ -27,15 +27,16 @@
 }
 
 impl UnsignedSubmitter {
-    pub(crate) fn new<Block, PBlock, PClient>(
+    pub(crate) fn new<Block, PBlock, PClient, Spawner>(
         primary_chain_client: Arc<PClient>,
-        spawner: Box<dyn SpawnNamed + Send + Sync>,
+        spawner: &Spawner,
     ) -> Self
     where
         Block: BlockT,
         PBlock: BlockT,
         PClient: HeaderBackend<PBlock> + ProvideRuntimeApi<PBlock> + 'static,
         PClient::Api: ExecutorApi<PBlock, Block::Hash>,
+        Spawner: SpawnNamed + ?Sized,
     {
         let (sender, mut receiver) = mpsc::channel(UNSIGNED_MESSAGE_BUFFER_SIZE);
```

liuchengxu replied:
Hmm, looks like the trick is using `spawner.as_ref()`, I was using `&spawer` when calling UnsignedSubmitted::new(), the compiler is not smart enough to lead me to the right place, thank you!

nazar-pc commented on cumulus/client/cirrus-executor/src/bundle_processor.rs:
```diff
@@ -626,32 +652,13 @@ where
+            .try_submit(UnsignedMessage::Fraud(fraud_proof))
+        {
+            let mut pending_unsigned = self.pending_unsigned.lock();
+            pending_unsigned.replace(msg);
```
Hm... but this will just drop existing pending extrinsic (if there is any), is that okay? I was expecting that the pipeline will pause until sending is possible again.

liuchengxu replied:
Only when no pending extrinsic can this line be reached, at the beginning of processing the bundles/gossiped messages, it will first check the pending extrinsic and try to resend, only continue the processing if the resend is okay when there is a pending extrinsic, e.g., https://github.com/subspace/subspace/pull/695/files#diff-de30ccb67696749ca21e43ded6b44ae00cd2822d3137191c39d830d80c2f81f7R534-R550. So the pipeline will pause if the message buffer is full as you expect.

nazar-pc replied:
Logic is too complex, I don't see how `on_bundle` there has anything to do with `submit_fraud_proof` method that is called from a completely different place. Even if that is correct, it is extremely not obvious to me.

liuchengxu replied:
I don't see any incorrectness in it, you need to check whether the data producer should continue the work at the beginning before it proceeds if you want to pause it when there are too many of them, `on_bundle` is one message source of them.

Replacing `pending_unsigned()` by checking if the channel still has some capacity should make it more obvious, and resolves the comment below.

nazar-pc commented on cumulus/client/cirrus-executor/src/bundle_processor.rs:
```diff
@@ -185,6 +193,24 @@ where
         maybe_new_runtime: Option<Cow<'static, [u8]>>,
     ) -> Result<(), sp_blockchain::Error> {
+        {
+            let mut pending_unsigned = self.pending_unsigned.lock();
```
I don't see the benefit of this `pending_unsigned`, it seems to effectively add `+1` to the capacity of the bounded channel in the `unsigned_submitter`.

liuchengxu replied:
Hmm, I intended to convey that the message producers(bundle_processor, gossiped_message_handler) can easily build their own additional buffer on top of the global one based on their priority, they're considered to be equivalently important for now though.

`pending_unsigned` can be replaced with checking if there is still some capacity in the channel and only continues when the channel still has some capacity, which is kind of conservative as not every `process_bundles` will be needed to submit this unsigned extrinsic. I have no strong opinion.

nazar-pc replied:
Why would they need to have their own buffer if they can't push messages into the global one anyway? Why not have multiple separate bounded channels instead of this one then? It seems to be overengineered and is hard to verify correctness of.

liuchengxu committed:
Replace `pending_unsigned` with a combo of `try_reserve()` and `send`

Now it first tries to reserve a slot before actually sending it, if the
submitter fails to reserve a slot, then the operation(bundle processing,
etc) will be paused.

Also applied another minor review suggestion and fixed `cargo doc`.

nazar-pc commented on cumulus/client/cirrus-executor/src/bundle_processor.rs:
```diff
@@ -185,6 +190,12 @@ where
         shuffling_seed: Randomness,
         maybe_new_runtime: Option<Cow<'static, [u8]>>,
     ) -> Result<(), sp_blockchain::Error> {
+        let fraud_proof_permit = self.unsigned_submitter.try_reserve().map_err(|_| {
+            sp_blockchain::Error::Application(Box::from(
+                "Channel for submitting unsigned extrinsics is full",
+            ))
+        })?;
```
Previous revision was returning `Ok` when channel is full, so this change isn't equivalent. Will some bundles be skipped entirely if this error happens and break the client? This is what seems to happen judging by the place where `process_bundles` is called from, I don't see any retry mechanism or anything like that, it just moves on with broken state.

liuchengxu replied:
The only difference from the previous version is that an error message of the channel being full will be logged, but now I realize that none of them is appropriate without the retry mechanism. The state won't be broken as the program now will panic if a primary block is skipped and the bundle processor attempts to process the next primary block due to the primary number check(see the assertation below `New secondary best number must be equal to the primary number`).

Okay, now the problem becomes more complicated, currently, we rely on the primary block import notification which is sort of one-off, for each imported primary block, the notification of that block will only be sent once and gone. When processing a primary block, if `process_bundles` returns an error, the executor is just broken at present, the state is not corrupted, but it can not move forward until the error is fixed and restarted.

Besides the retry mechanism, another problem begins to emerge: even though the retry mechanism exists, while the executor is retrying, the primary node is still syncing, which is a full node instead of an archival node and we won't expect the executor has to maintain a primary node keeping all the blocks, so if the retry takes too long, the primary node can proceed to a much higher block and prune the old blocks which are not yet processed by the executor, even when the executor retry is successful later, the corresponding primary block is unavailable and the executor get stuck there. I need some more time to think it over.

liuchengxu replied:
Rebased and added TODO which will be worked on in the nearest feature.

liuchengxu committed:
Add TODO of handling the error returned by `process_bundles()` properly

nazar-pc reviewed (commented):
Considering that we don't actually fix the problem and when we do, we'll pause block import pipeline, the whole `UnsignedSubmitter` seems to be unnecessary? With the same luck we could have just send unsigned extrinsics directly since we know corresponding processes are running sequentially and pause block import if necessary.

Also another test failed in Windows CI.
